### PR TITLE
enh(API): refactor DELETE platform/topology/id 

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR is a rework of DELETE platform/topology/id endpoint.
- Now when deleting a platform from a topology will also totally remove the platform (not only the topology)
- When a Remote is reconverted to a Central, Delete it from its previous parent.

**Fixes** # MON-6553

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x
- [x] 21.04.x (master)

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 1]
```diff
@@ -195,4 +218,115 @@ public function registerPlatformToParent(
             );
         }
     }
+
+    public function deletePlatformToParent(
```
Comment (kduret) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:222]: missing php doc

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 2]
```diff
@@ -195,4 +218,115 @@ public function registerPlatformToParent(
...
+        Platform $platform,
+        PlatformInformation $platformInformation,
+        ?Proxy $proxyService = null
```
Comment (kduret) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:225]: should be named `$proxy` cause it's an entity, not a service

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 3]
```diff
@@ -61,6 +68,74 @@ public function __construct(HttpClientInterface $httpClient, ApiPlatform $apiPla
...
+     *
+     * @param Platform $platform
+     * @param PlatformInformation $platformInformation
+     * @param Proxy $proxyService
+     * @return string
+     */
+    private function getToken(
+        Platform $platform,
+        PlatformInformation $platformInformation,
+        Proxy $proxyService = null
```
Comment (kduret) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:82]: should be named `$proxy` cause it's an entity, not a service

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 4]
```diff
@@ -61,6 +68,74 @@ public function __construct(HttpClientInterface $httpClient, ApiPlatform $apiPla
...
+    ): string {
+        // Central's API endpoints base path building
+        $this->baseApiEndpoint = $platformInformation->getApiScheme() . '://'
+            . $platformInformation->getCentralServerAddress() . ':'
+            . $platformInformation->getApiPort() . '/'
+            . $platformInformation->getApiPath() . '/api/v'
+            . ((string) $this->apiPlatform->getVersion()) . '/';
```
Comment (kduret) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:89]: IMHO, base endpoint should not contain trailing slash
It's like `__DIR__` constant, it does not add trailing slash

File: src/Centreon/Domain/MonitoringServer/MonitoringServerService.php [hunk 5]
```diff
@@ -130,4 +130,16 @@ public function notifyConfigurationChanged(MonitoringServer $monitoringServer):
...
+
+    /**
+     * @inheritDoc
+     */
+    public function deleteServer(int $monitoringServerId): void
+    {
+        try {
+            $this->monitoringServerRepository->deleteServer($monitoringServerId);
+        } catch (\Exception $ex) {
+            throw new MonitoringServerException('Error when deleting a monitoring server', 0, $ex);
```
Comment (kduret) [inline src/Centreon/Domain/MonitoringServer/MonitoringServerService.php:142]: COuld you add translation ?

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php (hunk 4 above)
Comment (jeremyjaouen) [reply src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php]: I'm not pretty sure if I understood well your feedback, I've changed he trailing slash for a DIRECTORY_SEPARATOR constant.

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php (hunk 2 above)
Comment (jeremyjaouen) [reply src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php]: indeed I also changed it into the other method RegisterPlatformToParent

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 6]
```diff
@@ -195,4 +218,118 @@ public function registerPlatformToParent(
...
+            $token = $this->getToken($platform, $platformInformation, $proxy);
+
+            $getPayload = [
+                'headers' => [
+                    "X-AUTH-TOKEN" => $token
+                ]
+            ];
+            $getResponse = $this->httpClient->request(
+                'GET',
+                $this->baseApiEndpoint . '/platform/topology/',
```
Comment (kduret) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:240]:
```suggestion
                $this->baseApiEndpoint . '/platform/topology',
```

**Review by kduret — DISMISSED** [review]

File: src/Centreon/Domain/RemoteServer/Interfaces/RemoteServerRepositoryFileInterface.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+/*
+ * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Domain/RemoteServer/Interfaces/RemoteServerRepositoryFileInterface.php:4]: 2021

File: src/Centreon/Domain/RemoteServer/Interfaces/RemoteServerRepositoryFileInterface.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Centreon\Domain\RemoteServer\Interfaces;
+
+interface RemoteServerRepositoryFileInterface
+{
+    /**
+     * Update the instance mode of centreon to Central.
```
Comment (sc979) [inline src/Centreon/Domain/RemoteServer/Interfaces/RemoteServerRepositoryFileInterface.php:29]:
```suggestion
     * Update the platform instance mode to Central.
```

File: src/Centreon/Domain/RemoteServer/Interfaces/RemoteServerRepositoryFileInterface.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    public function updateInstanceModeCentral(): void;
+
+    /**
+     * Update the instance mode of centreon to Remote.
```
Comment (sc979) [inline src/Centreon/Domain/RemoteServer/Interfaces/RemoteServerRepositoryFileInterface.php:34]:
```suggestion
     * Update the platform instance mode Remote.
```

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 10]
```diff
@@ -195,4 +218,118 @@ public function registerPlatformToParent(
...
+            $getResponse = $this->httpClient->request(
+                'GET',
+                $this->baseApiEndpoint . '/platform/topology',
+                $getPayload
+            );
+
+            // Get request status code and return the error message
+            if (Response::HTTP_OK !== $getResponse->getStatusCode()) {
+                $errorMessage = sprintf(
+                    _("The remote: '%s'@'%s' cannot be found to the Central"),
```
Comment (sc979) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:247]: translation missing
```suggestion
                    _("The remote: '%s'@'%s' cannot be found on the Central"),
```

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 11]
```diff
@@ -195,4 +218,118 @@ public function registerPlatformToParent(
...
+            ];
+            $getResponse = $this->httpClient->request(
+                'GET',
+                $this->baseApiEndpoint . '/platform/topology',
+                $getPayload
+            );
+
+            // Get request status code and return the error message
+            if (Response::HTTP_OK !== $getResponse->getStatusCode()) {
+                $errorMessage = sprintf(
```
Comment (sc979) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:246]: missing translation

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 12]
```diff
@@ -195,4 +218,118 @@ public function registerPlatformToParent(
...
+                throw new PlatformConflictException(
+                    $errorMessage
+                );
+            }
+
+            // Parse the response body to found the platform we want to delete
+            $responseBody = json_decode($getResponse->getContent(false), true);
+            $platformsOnParent = $responseBody['graph']['nodes'];
+
+            $platformToDeleteId = null;
```
Comment (sc979) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:266]:
```suggestion
            $platformIdToDelete = null;
```

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 13]
```diff
@@ -195,4 +218,118 @@ public function registerPlatformToParent(
...
+            foreach ($platformsOnParent as $topologyId => $platformOnParent) {
+                if ($platformOnParent['metadata']['address'] === $platform->getAddress()) {
+                    $platformToDeleteId = $topologyId;
+                }
+            }
+
+            if ($platformToDeleteId === null) {
+                throw new PlatformConflictException(_("The platform '%s'@'%s' can't be found on the Central."));
```
Comment (sc979) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:274]:
```suggestion
                throw new PlatformConflictException(_("The platform '%s'@'%s' cannot be found on the Central."));
```

Comment (sc979) [reply src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:274]: If you change the message on line 247, only one translation will be missing

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 14]
```diff
@@ -195,4 +218,118 @@ public function registerPlatformToParent(
...
+            $platformsOnParent = $responseBody['graph']['nodes'];
+
+            $platformToDeleteId = null;
+            foreach ($platformsOnParent as $topologyId => $platformOnParent) {
+                if ($platformOnParent['metadata']['address'] === $platform->getAddress()) {
+                    $platformToDeleteId = $topologyId;
+                }
+            }
+
+            if ($platformToDeleteId === null) {
```
Comment (sc979) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:273]: missing translation

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 15]
```diff
@@ -195,4 +218,118 @@ public function registerPlatformToParent(
...
+            $deleteResponse = $this->httpClient->request(
+                'DELETE',
+                $this->baseApiEndpoint . '/platform/topology/' . $platformToDeleteId,
+                $deletePayload
+            );
+
+            // Get request status code and return the error message
+            if (Response::HTTP_NO_CONTENT !== $deleteResponse->getStatusCode()) {
+                $errorMessage = sprintf(
+                    _("The remote: '%s'@'%s' cannot be delete from the Central"),
```
Comment (sc979) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:291]: missing translation

Are you sure that the error will be linked to the central, and not its children ?

File: config/packages/prod/Centreon.yaml [hunk 16]
```diff
@@ -253,10 +253,14 @@ services:
     Centreon\Domain\RemoteServer\Interfaces\RemoteServerServiceInterface:
         class: Centreon\Domain\RemoteServer\RemoteServerService
 
-    Centreon\Domain\RemoteServer\Interfaces\RemoteServerRepositoryInterface:
+    Centreon\Domain\RemoteServer\Interfaces\RemoteServerRepositoryFileInterface:
```
Comment (callapa) [inline config/packages/prod/Centreon.yaml:256]: The name of an interface should remain abstract and should not provide technical information. The term 'File' is a technical indication.

File: src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php [hunk 17]
```diff
@@ -61,72 +68,88 @@ public function __construct(HttpClientInterface $httpClient, ApiPlatform $apiPla
...
     }
 
+    /**
+     * Get a valid token to request the API.
+     *
+     * @param Platform $platform
+     * @param PlatformInformation $platformInformation
+     * @param Proxy $proxy
+     * @return string
+     */
```
Comment (callapa) [inline src/Centreon/Infrastructure/PlatformTopology/PlatformTopologyRegisterRepositoryAPI.php:78]: You forgot to add PHPDoc regarding exceptions

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by kduret — DISMISSED** [review]

File: src/Centreon/Domain/RemoteServer/RemoteServerService.php [hunk 18]
```diff
@@ -168,8 +173,34 @@ public function convertCentralToRemote(PlatformInformation $platformInformation)
...
+        $platform = $this->platformTopologyRepository->findTopLevelPlatform();
+        $this->platformTopologyRegisterRepository->deletePlatformToParent(
+            $platform,
+            $platformInformation,
+            $this->proxyService->getProxy()
+        );
+
+        /**
+         * Find any children platform and remove them,
+         * as they now attached to the Central and no longer to this platform.
```
Comment (sc979) [inline src/Centreon/Domain/RemoteServer/RemoteServerService.php:191]:
```suggestion
         * as they are now attached to the Central and no longer to this platform.
```

**Review by sc979 — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Review by loiclau — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**